### PR TITLE
Develop/drgn 10904 create big lock

### PR DESCRIPTION
**Pull request by omkar3141**

### CODE REVIEW

File: subsys/bluetooth/controller/Kconfig [hunk 1]
```diff
@@ -20,6 +20,38 @@ config BLECTLR_SIGNAL_STACK_SIZE
 	  Size of the signal handler thread stack, used to process lower
 	  priority signals in the controller.
 
+comment "Multithreading behavior of BLE controller APIs"
+
+config BLECTLR_SHARED_API_BLOCKING
```
Comment (PavelVPV) [inline subsys/bluetooth/controller/Kconfig]: I think it can be `menuconfig` instead of `config`. Then all `_API_BLOCKING` options could be hidden inside separate page in menuconfig

File: subsys/bluetooth/controller/CMakeLists.txt [hunk 2]
```diff
@@ -7,6 +7,7 @@
 zephyr_library()
 
 zephyr_library_sources(
+  lock.c
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/CMakeLists.txt]: Rename file to multithreading_lock.c?

File: subsys/bluetooth/controller/crypto.c [hunk 3]
```diff
@@ -11,56 +11,65 @@
 #include <soc.h>
 #include <logging/log.h>
 
-#include "nrf_error.h"
+#include <blectlr_util.h>
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/crypto.c]: Please do an upmerge. blectlr_util.h does no longer exist

File: subsys/bluetooth/controller/crypto.c [hunk 4]
```diff
@@ -11,56 +11,65 @@
...
+#include <ble_controller_soc.h>
 
-#include "ble_controller_soc.h"
-#include "blectlr_util.h"
+#include "nrf_error.h"
+#include "lock.h"
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/crypto.c]: Seems like lock.h is missing from the pull request

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+#ifndef LOCK_H__
+#define LOCK_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr.h>
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.h:20]: Are these includes required?

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+#include <soc.h>
+#include <gpio.h>
+
+/** Macro representation for unlock API */
+#define API_UNLOCK api_unlock
+
+/** Macro representation for typical usage of locking API */
+#define API_LOCK_AND_RETURN_ON_FAIL(err) \
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/multithreading_lock.h]: Why should the API take in a variable here? It creates it's own scope in the do-while loop anyway.

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/**
+ * @lock.h
+ *
+ * @brief This file defines APIs needed to lock the BLE controller APIs for threadsafe operation.
+ */
+
+#ifndef LOCK_H__
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/multithreading_lock.h]: This guard should match the name of the file

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ * will wait for timeout with a given timeout value specified by
+ * @ref CONFIG_BLECTLR_SHARED_API_BLOCKING_WITH_TIMEOUT and return error if lock cannot be
+ * obtained. If @ref CONFIG_BLECTLR_SHARED_API_BLOCKING is not specified, API will return
+ * immediately in the event of not being able to obtain a lock.
+ *
+ * @retval 0			Success
+ * @retval - ::NRF_EBUSY	Returned without waiting.
+ * @retval - ::NRF_EAGAIN	Waiting period timed out.
+ */
+int api_lock(void);
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/multithreading_lock.h]: The API and file/namespace should match?

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 6 above)
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.h]: The macro should not do anything if multithreading is turned off

File: subsys/bluetooth/controller/Kconfig [hunk 9]
```diff
@@ -20,6 +20,41 @@ config BLECTLR_SIGNAL_STACK_SIZE
 	  Size of the signal handler thread stack, used to process lower
 	  priority signals in the controller.
 
+comment "Multithreading behavior of BLE controller APIs"
+
+menuconfig BLECTLR_SHARED_API_BLOCKING
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/Kconfig]: BLECTLR_SUPPORT_MULTITHREADING?

File: subsys/bluetooth/controller/multithreading_lock.c [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "multithreading_lock.h"
+
+#include <soc.h>
+#include <gpio.h>
+
+/* Semaphore with a single count functioning as a lock. */
+K_SEM_DEFINE(ble_controller_lock, 1, 1);
+
+int api_lock(void)
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.c]: I don't think it is necessary to define these functions

File: subsys/bluetooth/controller/Kconfig (hunk 9 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/controller/Kconfig]: Or BLCTLR_THREADSAFE

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+/** Macro representation for typical usage of locking API */
+#define API_LOCK_AND_RETURN_ON_FAIL(err) \
+	do {				 \
+		err = api_lock();	 \
+		if (err != 0) {		 \
+			return err;	 \
+		}			 \
+	} while (0);
+
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.h:52]: What about the following:
```
#ifdef BLECTLR_CONFIG_THREADSAFE
#define THREADSAFE_CALL(api_call) \
 do {
    api_lock();
    api_call();
    api_unlock();
  while (0);
#else
THREADSAFE_CALL(api_call) api_call
#endif
```
(error code return is not included in this example)

Comment (omkar3141) [reply subsys/bluetooth/controller/multithreading_lock.h:52]: That will not work. Most api_calls() have return values and depending upon the returns driver code will work differently.

Comment (PavelVPV) [reply subsys/bluetooth/controller/multithreading_lock.h:52]: The error code can be wrapped as well, can't it?

File: subsys/bluetooth/controller/Kconfig [hunk 12]
```diff
@@ -29,4 +29,40 @@ choice BLE_CONTROLLER_VARIANT
 
 endchoice
 
+
+comment "Thread safety behavior of BLE controller APIs"
+
+menuconfig BLECTLR_THREADSAFE_BLOCKING
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/Kconfig]: BLECTLR_THREADSAFE

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 5 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/controller/multithreading_lock.h:20]: Do you need to include <zephyr.h>?

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 11 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/controller/multithreading_lock.h:52]: I think you should look at this once more

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 5 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/multithreading_lock.h:20]: Required for type "s32_t" (zephyr/types.h) in the macro definition. Also for IS_ENABLED (misc/util.h). 
Both of these are included via `zephyr.h`.

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 11 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/multithreading_lock.h:52]: Discussed offline.

**Review by thomasstenersen — CHANGES_REQUESTED** [review]
I'm commenting here to avoid going into the various details before addressing the bigger picture. In general, the design of the `multithreading_lock.h` interface seems to invite poor usage patterns. It is also breaking the Linux kernel coding guidelines:

<https://www.kernel.org/doc/html/latest/process/coding-style.html#macros-enums-and-rtl>

> 1.  Macros that affect control flow is a very bad idea. It looks like a function call but exits the calling function; don’t break the internal parsers of those who will read the code.

For example, I have allocated a slab of memory I cannot call `API_LOCK_AND_RETURN_ON_FAIL` -- it will cause a memory leak.

Furthermore, the API seems to define multiple different ways of achieving the same thing, let's improve that. By using a single variable, we may support all use cases.

`CONFIG_BLECTRL_THREADSAFE_TIMEOUT_MSEC` can take 3 kinds of values:

-   `K_FOREVER: -1` &#x2013; the API is (forever) blocking
-   `K_NO_WAIT: 0` &#x2013; the API is nonblocking
-   `K_MSEC(n): n > 0` &#x2013; the API blocks for a certain timeout

As I see it, there are at least three general approaches here:

```C

/* Option 1 */
#define MULTITHREAD_LOCK_API_CALL(ERROR_CODE, FUNCALL)                  \
    do {                                                                \
    ERROR_CODE = multithread_api_lock(CONFIG_BLECTRL_THREADSAFE_TIMEOUT_MSEC); \
    if (!ERROR_CODE) {                                                  \
        FUNCALL;                                                        \
        multithread_api_unlock();                                       \
    } while (0)

/* Option 2 */
#define MULTITHREAD_LOCK_AQUIRE()               \
    multithread_api_lock(CONFIG_BLECTRL_THREADSAFE_TIMEOUT_MSEC);

#define MULTITHREAD_LOCK_RELEASE()              \
    multithread_api_unlock();


/* Option 3 */
#define MULTITHREAD_LOCK(ERROR_CODE)                                    \
	for (int i = 0;                                                     \
	     i < 1 && (ERROR_CODE = multithread_api_lock(CONFIG_BLECTRL_THREADSAFE_TIMEOUT_MSEC)) == 0; \
	     multithread_api_unlock(), ++i)

/* Usage: */
int option1(void) {
    int error_code = 0;
    MULTITHREAD_LOCK_API_CALL(error_code, unsafe_api(1, 2, 3));
    return error_code;
}

int option2(void) {
    int error_code;
    error_code = MULTITHREAD_LOCK_AQUIRE();
    if (!error_code) {
        error_code = unsafe_api(1, 2, 3);
    }
    MULTITHREAD_LOCK_RELEASE();
    return error_code;
}

int option3(void) {
    int error_code;
    MULTITHREAD_LOCK(error_code) {
        error_code = unsafe_api(1, 2, 3);
    }
    return error_code;
}
```

I prefer option 3, though option 2 is more traditional. Option 1 could be tricky when there are functions with different types of return signatures. Option 3 has the potential for bugs if the braces are missed, but so does every `if/while/for`.

**Review by omkar3141 — COMMENTED** [review]
I would not prefer option 3 as it has implicit `for` and it is creating a trick that should be remembered. If not, possibly spend time in debugging API calls that may not be easy to debug. Implications for missing braces of `if/while/for` are much more easy to catch as these constructs are used directly.

I find option 2 is more straight forward and less error-prone.

**Comment by thomasstenersen** [comment]
A rename of that `#define` would make it clearer that it is intended to be used with the braces. E.g., `WITH_MULTITHREAD_LOCK(..)` (though that might not follow API convention @rugeGerritsen?). It is not a widely used pattern, so you can decide not to use it. Option 2 is equally (if not more) unsafe due to the potential unhandled return value. I don't know how strict the default compilation settings in `fw-nrfconnect-nrf` are set.

File: subsys/bluetooth/controller/crypto.c [hunk 13]
```diff
@@ -11,56 +11,64 @@
...
+#include "multithreading_lock.h"
 
 #define LOG_MODULE_NAME ble_controller_crypto
 LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
 
-
 int bt_rand(void *buf, size_t len)
 {
-	// @todo: support CONFIG_MULTITHREADING
+	s32_t errcode;
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/crypto.c]: I see this multiple places, any reason for using `s32_t` when the API is using a plain `int`?

File: subsys/bluetooth/controller/power_clock_driver.c [hunk 14]
```diff
@@ -10,20 +10,53 @@
...
+
+	return errcode;
+}
 
 static int hf_clock_start(struct device *dev, clock_control_subsys_t sub_system)
 {
 	ARG_UNUSED(dev);
 
-	if (ble_controller_hf_clock_request(NULL) != 0) {
+	if (ble_controller_hf_clock_request_wlock(NULL) != 0) {
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/power_clock_driver.c]: this `_wlock()` pattern appears several places where more-or-less the same function is defined again. Consider generalizing this to remove all the copy-pasted code.

**Comment by omkar3141** [comment]
Irrespective of compilation settings, option 2 will create an execution block if the return value is ignored, and intern it could be caught easily. If we can add custom compilation settings in nrf, it will be more useful.

**Comment by thomasstenersen** [comment]
That's not true. If the lock fails, the code should not continue to execute.

**Comment by PavelVPV** [comment]
Option 3 won't work if you use return inside the block. But you can still use return in the middle of the code with other patterns.

**Comment by thomasstenersen** [comment]
True, @PavelVPV, I suppose there are both advantages and disadvantages. This is where we should've had C++ constructors and destructors. Anywho, there will be a drawback with any one of these. I simply want to optimize for reduction of bugs, minimize copy-paste patterns and the readability of the code using these primitives.

File: subsys/bluetooth/controller/power_clock_driver.c (hunk 14 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/power_clock_driver.c]: Removed these and updated the code as appropriate.

**Comment by PavelVPV** [comment]
With Option 1 we don't know which function returned the error code: the lock function or the "locked" one. It also mixes Zephyr's error codes with the "locked" API calls.

So I would prefer the Option 2. 

@thomasstenersen I see your point. Therefore, I think, that among all these options, the 2nd one is more flexible intuitively understandable pattern, where it is easier to catch an error.

**Comment by omkar3141** [comment]
Option 2 gives rise to significant manual code repetition everywhere the lock API is used. I found option 1 is giving more compact usage and gives user an option to completely exclude the locking if desired. Also, from the debug perspective it should be possible to debug this implementation using either logging or using pin toggles.

File: subsys/bluetooth/controller/hci_driver.c [hunk 15]
```diff
@@ -40,7 +41,10 @@ void blectlr_assertion_handler(const char *const file, const u32_t line)
 
 static int cmd_handle(struct net_buf *cmd)
 {
-	if (hci_cmd_put(cmd->data)) {
+	int errcode;
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/hci_driver.c]: errcode should be int32_t. Similar for other occurrences

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 16]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/**
+ * @lock.h
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.h]: Documentation is outdated

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr.h>
+
+/** Macro for calling ble controller API with a return value in threadsafe manner */
+#if IS_ENABLED(CONFIG_BLECTLR_THREADSAFE_BLOCKING)
+#define THREADSAFE_CALL_WITH_RETCODE(ERROR_CODE_VAR, FUNC_CALL)                 \
+	do {                                                                   \
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.h]: Please align the backslashes

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+#else
+#define THREADSAFE_CALL_WITH_RETCODE(ERROR_CODE_VAR, FUNC_CALL)                 \
+	do {                                                                   \
+		ERROR_CODE_VAR = FUNC_CALL;                                    \
+	} while (0)
+#endif
+
+/** Macro for calling ble controller API without a return value in threadsafe manner */
+#if IS_ENABLED(CONFIG_BLECTLR_THREADSAFE_BLOCKING)
+#define THREADSAFE_CALL_WITHOUT_RETCODE(ERROR_CODE_VAR, FUNC_CALL)              \
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.h]: Is this used?

File: subsys/bluetooth/controller/hci_driver.c (hunk 15 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/hci_driver.c]: This was pointed by @thomasstenersen , and hence I used the same datatype for the errorcode  as that that of the return type of the function.

Is there any guideline on this? Which datatype sets to use in the NCS? 
For ex: int (stdint) vs s32_t (zephyr) vs (int32_t) Nordic?

**Comment by thomasstenersen** [comment]
Option 2 is the traditional one, probably for a reason, so that's what I'd vote for.

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 18 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/multithreading_lock.h]: Not currently. But it is provided if we have the use case in future.

Comment (rugeGerritsen) [reply subsys/bluetooth/controller/multithreading_lock.h]: Because it is not used it should be removed

**Comment by rugeGerritsen** [comment]
We seem to agree on using option 2

File: subsys/bluetooth/controller/hci_driver.c [hunk 19]
```diff
@@ -167,18 +178,35 @@ static void recv_thread(void *p1, void *p2, void *p3)
...
 
-		while (hci_data_get(hci_buffer) == 0) {
-			data_packet_process(hci_buffer);
-		}
-
-		while (hci_evt_get(hci_buffer) == 0) {
-			event_packet_process(hci_buffer);
-		}
+		while (1) {
+			if (!(errcode = MULTITHREADING_LOCK_ACQUIRE())) {
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/hci_driver.c]: Please move the expression out of the if statement

File: subsys/bluetooth/controller/Kconfig [hunk 20]
```diff
@@ -29,4 +29,24 @@ choice BLE_CONTROLLER_VARIANT
 
 endchoice
 
+
+comment "Thread safety behavior of BLE controller APIs"
+
+config BLECTLR_THREADSAFE_BLOCKING
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/Kconfig]: BLOCKING is misleading. Please change

File: subsys/bluetooth/controller/multithreading_lock.c [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+int32_t multithreading_lock_acquire(int32_t timeout)
+{
+#if IS_ENABLED(CONFIG_BLECTLR_THREADSAFE_BLOCKING)
+	return k_sem_take(&ble_controller_lock, K_MSEC(timeout));
+#else
+	return 0;
+#endif
+}
+
+int32_t multithreading_lock_acquire_try(void)
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/multithreading_lock.c]: Do we really need this API?

Comment (omkar3141) [reply subsys/bluetooth/controller/multithreading_lock.c]: We can remove this for now and add it later if required.

File: subsys/bluetooth/controller/Kconfig (hunk 20 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/Kconfig]: changing 
BLECTLR_THREADSAFE_BLOCKING -> BLECTLR_THREADSAFETY and 
BLECTLR_THREADSAFE_BLOCKING_TIMEOUT -> BLECTLR_THREADSAFETY_LOCK_TIMEOUT

File: subsys/bluetooth/controller/hci_driver.c [hunk 22]
```diff
@@ -328,6 +373,7 @@ static int hci_driver_init(struct device *unused)
 	IRQ_DIRECT_CONNECT(NRF5_IRQ_TIMER0_IRQn, 0,
 			   ble_controller_TIMER0_IRQHandler, IRQ_ZERO_LATENCY);
 	IRQ_CONNECT(NRF5_IRQ_SWI5_IRQn, 4, SIGNALLING_Handler, NULL, 0);
+	IRQ_CONNECT(NRF5_IRQ_RNG_IRQn, 4, ble_controller_RNG_IRQHandler, NULL, 0);
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/hci_driver.c]: This should be removed as it is moved to the rng driver

File: subsys/bluetooth/controller/multithreading_lock.h [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+#else
+#define MULTITHREADING_LOCK_RELEASE()
+#endif
+
+/** @brief Try to take the lock while maintaining the specified blocking behavior.
+ *
+ * This API call will be blocked for the time specified by @ref
+ * CONFIG_BLECTLR_THREADSAFETY_LOCK_TIMEOUT and then return error code.
+ *
+ * @param  timeout		Timeout value for locking API.
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/multithreading_lock.h]: Should probably provide a unit to this. Rename the parameter to `timeout_msec` and put the unit in the description?

File: subsys/bluetooth/controller/power_clock_driver.c [hunk 24]
```diff
@@ -10,20 +10,32 @@
...
 	ARG_UNUSED(dev);
 
-	if (ble_controller_hf_clock_request(NULL) != 0) {
+	int32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+	if (errcode == 0) {
+		errcode = ble_controller_hf_clock_request(NULL);
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	if (errcode != 0) {
 		return -EFAULT;
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/power_clock_driver.c:25]: This is overriding the potential error code from `MULTITHREADING_LOCK_AQUIRE()` is that intentional?

File: subsys/bluetooth/controller/power_clock_driver.c [hunk 25]
```diff
@@ -10,20 +10,32 @@
...
 		while (!is_running) {
-			if (ble_controller_hf_clock_is_running(&is_running) != 0) {
+			errcode = MULTITHREADING_LOCK_ACQUIRE();
+			if (errcode == 0) {
+				errcode = ble_controller_hf_clock_is_running(
+					&is_running);
+				MULTITHREADING_LOCK_RELEASE();
+			}
+			if (errcode != 0) {
 				return -EFAULT;
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/power_clock_driver.c:39]: I see this several places in this file.

File: subsys/bluetooth/controller/rng_driver.c [hunk 26]
```diff
@@ -31,11 +33,18 @@ static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
...
 	u16_t bytes_left = len;
 	while (bytes_left > 0) {
-		int32_t bytes_read;
-		while ((bytes_read = ble_controller_rand_vector_get(p_dst, bytes_left)) <= 0) {
-			/* Put the thread on wait until next interrupt to get more
-			 * random values. */
-			k_sem_take(&rng_data.sem_sync, K_FOREVER);
+		int32_t bytes_read = 0;
+		while (bytes_read <= 0) {
+			if (!MULTITHREADING_LOCK_ACQUIRE()) {
```
Comment (thomasstenersen) [inline subsys/bluetooth/controller/rng_driver.c]: .. this reads "if not multithreading lock aquire" :disappointed: `== 0` for clarity? The next reader of this file will thank you :+1:

File: subsys/bluetooth/controller/power_clock_driver.c (hunk 24 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/power_clock_driver.c:25]: Yes. Even in the earlier case, this driver was returning -EFAULT when API call was unsuccessful. These changes keep the same behaviour. Probably that is how the driver interface works. Returning some other error code might break the user of this driver.

File: subsys/bluetooth/controller/power_clock_driver.c (hunk 25 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/power_clock_driver.c:39]: I am just maintaining error codes. You can have a look at the base version. API error code is ignored and some specific error code is returned.

File: subsys/bluetooth/controller/rng_driver.c (hunk 26 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/rng_driver.c]: In various files, I have tried to maintain the same style as that of the old style. Everywhere I see, a pattern like `if (!err)` ... `if (!some_call())` is used. I am continuing the same pattern. 

In the light of this comment, should we document such patterns somewhere and stick to the agreed convention? I will change it accordingly.

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 23 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/multithreading_lock.h]: I had done that earlier (something like `timeout_ms`) but changed my mind. I have followed the convention used everywhere in the zephyr/nrfxlib/nrf repos. In this case, I will prefer to keep the same convention. I will change the description though.

File: subsys/bluetooth/controller/rng_driver.c (hunk 26 above)
Comment (PavelVPV) [inline subsys/bluetooth/controller/rng_driver.c]: One thing I'm concerned about is that a thread, which called this function, can be blocked forever if it failed to acquire the lock and semaphore is nevere given (beacuse IRQ is disabled). So the thread will stuck in k_sem_take().

How about this code:

> while (MULTITHREADING_LOCK_ACQUIRE()) {
>    k_yield();
> }
> bytes_read = ble_controller_rand_vector_get(p_dst, bytes_left);
> MULTITHREADING_LOCK_RELEASE();
> k_sem_take(&rng_data.sem_sync, K_FOREVER);

File: subsys/bluetooth/controller/multithreading_lock.h (hunk 23 above)
Comment (thomasstenersen) [reply subsys/bluetooth/controller/multithreading_lock.h]: Ok, as long as the units are there in the description.

File: subsys/bluetooth/controller/rng_driver.c (hunk 26 above)
Comment (thomasstenersen) [reply subsys/bluetooth/controller/rng_driver.c]: Whether it should be `if (!err)` or `if (err==0)` is _not_ what I am arguing here. I am talking about how you'd read _this specific if-statement_. With the name of the function/macro, not knowing anything about what it returns, this code reads as if you're not able to acquire the lock, you should continue ;)

I'm simply asking you if you have considered this code from the point of view of someone reading this 4 weeks from now. Would you need to open `multithreading_lock.h` and read the definition of the function to understand what's going on?

Comment (omkar3141) [reply subsys/bluetooth/controller/rng_driver.c]: Not really. In case of Forever/Timeout blocking,`k_sem_take()` (which is called by the acquire function) will automatically put the calling thread in pending mode if semaphore cannot be taken.

Comment (PavelVPV) [reply subsys/bluetooth/controller/rng_driver.c]: After offline discussion it was decided to add one more function called MULTITHREADING_LOCK_ACQUIRE_FOREVER() to be used inside the RNG driver, because of the implementation of ble_controller and how this API is used by applications.

So this function will be used here, in rng_driver_entropy_get(), and in rng_driver_entropy_get_isr() for busywait case.

Comment (omkar3141) [reply subsys/bluetooth/controller/rng_driver.c]: alright.

**Review by thomasstenersen — APPROVED** [review]

File: subsys/bluetooth/controller/crypto.c [hunk 27]
```diff
@@ -10,19 +10,17 @@
...
+#include "multithreading_lock.h"
 
 #define LOG_MODULE_NAME ble_controller_crypto
 LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
 
-
 int bt_rand(void *buf, size_t len)
 {
-	if (buf == NULL) {
+	if (len > UINT8_MAX || buf == NULL) {
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/crypto.c]: Why did you change bt_rand?

File: subsys/bluetooth/controller/crypto.c [hunk 28]
```diff
@@ -46,32 +44,40 @@ int bt_rand(void *buf, size_t len)
...
 	LOG_HEXDUMP_DBG(plaintext, 16, "plaintext");
 
-	uint32_t errcode = ble_controller_ecb_block_encrypt(key, plaintext, enc_data);
+	int32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+	if (!errcode) {
+		errcode = ble_controller_ecb_block_encrypt(key, plaintext, enc_data);
+		MULTITHREADING_LOCK_RELEASE();
+	}
 
 	LOG_HEXDUMP_DBG(enc_data, 16, "enc_data");
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/crypto.c]: The `LOG_HEXDUMP_DBG` should be moved inside `if (!errcode)`

Comment (rugeGerritsen) [reply subsys/bluetooth/controller/crypto.c]: The same applies in other places

File: subsys/bluetooth/controller/crypto.c (hunk 27 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/crypto.c]: This is probably leftover after @PavelVPV changes in his PR. This condition should probably be removed now this condition was valid when ble_controller API was used directly.

**Review by PavelVPV — APPROVED** [review]

File: subsys/bluetooth/controller/crypto.c (hunk 28 above)
Comment (omkar3141) [reply subsys/bluetooth/controller/crypto.c]: Fixed.

File: subsys/bluetooth/controller/rng_driver.c [hunk 29]
```diff
@@ -31,11 +33,23 @@ static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
...
 	u16_t bytes_left = len;
 	while (bytes_left > 0) {
-		int32_t bytes_read;
-		while ((bytes_read = ble_controller_rand_vector_get(p_dst, bytes_left)) <= 0) {
-			/* Put the thread on wait until next interrupt to get more
-			 * random values. */
-			k_sem_take(&rng_data.sem_sync, K_FOREVER);
+		int32_t bytes_read = 0;
+		while (bytes_read <= 0) {
+			int32_t errcode = MULTITHREADING_LOCK_ACQUIRE_FOREVER_WAIT();
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/rng_driver.c]: Why is FOREVER_WAIT needed here?

Comment (omkar3141) [reply subsys/bluetooth/controller/rng_driver.c]: Decided to use regular ACQUIRE macro for rng_driver_get_entropy() and return immediately on error to prevent lock out.

**Review by rugeGerritsen — APPROVED** [review]
